unit_tests/test_NNLS.cpp

```diff
@@ -0,0 +1,240 @@
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+#ifdef CAROM_HAS_GTEST
+
+#include<gtest/gtest.h>
+#include "linalg/NNLS.h"
+#include <algorithm>
+#include <cmath>
+#include <cstdio>
+#include <cstring> // for memcpy
+#include <random>
+#include "mpi.h"
+#include "utils/mpi_utils.h"
+
+/**
+ * Simple smoke test to make sure Google Test is properly linked
+ */
+TEST(GoogleTestFramework, GoogleTestFrameworkFound) {
+    SUCCEED();
+}
+
+TEST(NNLS, solve_with_LINF)
+{
+    int nproc;
+    int rank;
+    MPI_Comm_size(MPI_COMM_WORLD, &nproc);
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    const int nrow = 300;
+    const int ncol = 1000;
+    const int ncol_local = CAROM::split_dimension(ncol);
+    std::vector<int> row_offset(nproc + 1);
+    const int total_cols = CAROM::get_global_offsets(ncol_local, row_offset,
+                           MPI_COMM_WORLD);
+    const double rel_tol = 0.05;
+    const double nnls_tol = 1.0e-11;
+
+    std::default_random_engine generator;
+    generator.seed(
+        1234); // fix the seed to keep the same result for different nproc.
+    std::uniform_real_distribution<> uniform_distribution(0.0, 1.0);
+    std::normal_distribution<double> normal_distribution(0.0, 1.0);
+
+    // distribute from a global matrix to keep the same system for different nproc.
+    CAROM::Matrix Gt(ncol, nrow, false);
+    for (int i = 0; i < ncol; i++)
+        for (int j = 0; j < nrow; j++)
+            Gt(i, j) = normal_distribution(generator);
+    Gt.distribute(ncol_local);
+
+    CAROM::Vector fom_sol(ncol_local, true);
+    CAROM::Vector rom_sol(ncol_local, true);
+    CAROM::Vector rhs(nrow, false);
+
+    // distribute from a global matrix to keep the same system for different nproc.
+    CAROM::Vector fom_sol_serial(ncol, false);
+    for (int c = 0; c < ncol; c++)
+        fom_sol_serial(c) = uniform_distribution(generator);
+    for (int c = 0; c < ncol_local; c++)
+        fom_sol(c) = fom_sol_serial(row_offset[rank] + c);
+
+    Gt.transposeMult(fom_sol, rhs);
+    rom_sol = 0.0;
+
+    CAROM::Vector rhs_lb(rhs);
+    CAROM::Vector rhs_ub(rhs);
+
+    for (int i = 0; i < rhs.dim(); ++i)
+    {
+        double delta = rel_tol * abs(rhs(i));
+        rhs_lb(i) -= delta;
+        rhs_ub(i) += delta;
+    }
+
+    CAROM::NNLSSolver nnls(nnls_tol, 0, 0, 2);
+    nnls.solve_parallel_with_scalapack(Gt, rhs_lb, rhs_ub, rom_sol);
+
+    int nnz = 0;
+    for (int i = 0; i < rom_sol.dim(); ++i)
+    {
+        if (rom_sol(i) != 0.0)
+        {
+            nnz++;
+        }
+    }
+
+    std::cout << rank << ": Number of nonzeros in NNLS solution: " << nnz
+              << ", out of " << rom_sol.dim() << std::endl;
+
+    MPI_Allreduce(MPI_IN_PLACE, &nnz, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
+
+    if (rank == 0)
+        std::cout << "Global number of nonzeros in NNLS solution: " << nnz << std::endl;
```

Comment:
Should there be a check that `nnz` is less than some number based on the result of this test?

Reply:
Do you mean a prescribed answer based on the given random seed? Yeha I think it's possible.

Reply:
Yes, I think the test would be stronger if it also checks `nnz`, and it would be necessary to fix the random seed to make the test deterministic.